# Day 1 · 산업재해 데이터 → Pandas → Embedding → PostgreSQL/pgvector

이 노트북의 목표는 **검색 가능한 문서를 만들어 DB에 저장하고, 실제로 한 건 조회하는 것**입니다. Agent, LLM 답변 생성, Streamlit은 다음 날 작업입니다.

실행 위치: WSL의 `rag-project-1/src/rag_project_1/Project_1`에 있는 이 노트북. 커널은 `rag-project-1/.venv/bin/python`을 선택하세요. 저장소 루트에서 Jupyter를 실행해도 아래 경로 찾기 셀이 동작합니다. 위에서 아래로 순서대로 실행합니다. 임베딩 API는 호출량에 따라 비용이 발생하므로 처음에는 각 자료 유형에서 고른 `MAX_DOCUMENTS = 20`건으로 검증하고, 전체 적재 시 `None`으로 바꾸세요. 같은 `source_id`는 재실행 시 갱신됩니다.

**준비물**: 저장소 루트 `.env`의 `OPENAI_API_KEY`, 실행 중인 PostgreSQL의 pgvector 확장. DB 연결은 `DATABASE_URL` 환경 변수가 있으면 우선 사용하고, 없으면 이 저장소의 `docker-compose.yaml`에 맞춘 로컬 접속값을 사용합니다. `CREATE EXTENSION vector` 권한이 필요합니다. 접속 정보는 출력하지 않습니다.


## 1. 경로와 데이터 목록

기존 `Day1_데이터분석.ipynb`와 같이 SIF 엑셀은 `header=2`로 읽습니다. 파일명 패턴으로 찾되 기대 파일이 정확히 하나인지 검사합니다.

| 파일 | 사용하는 컬럼 | 문서 단위 |
|---|---|---|
| SIF 아카이브 엑셀, `아카이브(제조업등)` | `연번`, `산재업종\n(대분류/중분류/소분류)`, `재해개요`, `기인물`, `고위험작업·상황`, `재해유발요인`, `위험성 감소대책(예시)` | 사례 1행 = 1문서 |
| 같은 엑셀, `아카이브(건설업)` | `연번`, `고위험작업·상황`, `Unnamed: 3`, `Unnamed: 4`, `재해종류`, `재해개요`, `기인물`, `재해유발요인`, `위험성 감소대책(예시)` | 사례 1행 = 1문서. `Unnamed: 3/4`는 중·소 작업분류 |
| `산업중분류별 규모별 사고사망자수` CSV | `대업종`, `구분`, 나머지 10개 사업장 규모별 수치 | 업종 1행 = 1문서 |
| `산업중분류별 규모별 사고재해자수` CSV | 위와 동일 | 업종 1행 = 1문서 |
| `산업중분류별 규모별 사망만인율` CSV | `대업종`, `구분`, 나머지 10개 규모별 비율 | 업종 1행 = 1문서 |
| `산업중분류별 규모별 사업장수` CSV | `대업종`, `구분`, 나머지 10개 규모별 수치 | 업종 1행 = 1문서 |

CSV의 개별 규모 컬럼명은 파일마다 다르므로 아래에서 실제 헤더를 출력하고, 앞의 두 식별 컬럼을 제외한 10개를 사용합니다. 원본의 사고사망자수·사고재해자수 CSV에도 규모 컬럼명이 `근로자수`로 되어 있으므로 이 노트북은 원본 표기를 그대로 보존하고, 지표 종류는 파일명으로 구분합니다. `Unnamed: 0`은 비어 있는 엑셀 서식 컬럼이어서 사용하지 않습니다.


In [1]:
from pathlib import Path
import os
import re

import pandas as pd
from dotenv import load_dotenv

# 커널이 노트북 폴더나 저장소 루트에서 시작해도 같은 data를 찾습니다.
cwd = Path.cwd().resolve()
project_candidates = [cwd, *(parent / 'src/rag_project_1/Project_1' for parent in (cwd, *cwd.parents))]
PROJECT_DIR = next((candidate for candidate in project_candidates if (candidate / 'data').is_dir() and candidate.name == 'Project_1'), None)
if PROJECT_DIR is None:
    raise FileNotFoundError('rag-project-1 저장소 또는 src/rag_project_1/Project_1 안에서 Jupyter를 실행하세요.')
REPO_ROOT = PROJECT_DIR.parents[2]
print('프로젝트:', PROJECT_DIR)
print('저장소:', REPO_ROOT)
load_dotenv(REPO_ROOT / '.env')
DATA_DIR = PROJECT_DIR / 'data'

def one_file(pattern):
    matches = list(DATA_DIR.glob(pattern))
    if len(matches) != 1:
        raise FileNotFoundError(f'{pattern}: 파일이 정확히 1개여야 합니다. 발견: {len(matches)}')
    return matches[0]

sif_path = one_file('*SIF*아카이브*.xlsx')
stat_patterns = {
    '사고사망자수': '*산업중분류별 규모별 사고사망자수*.csv',
    '사고재해자수': '*산업중분류별 규모별 사고재해자수*.csv',
    '사망만인율': '*산업중분류별 규모별 사망만인율*.csv',
    '사업장수': '*산업중분류별 규모별 사업장수*.csv',
}
stat_paths = {name: one_file(pattern) for name, pattern in stat_patterns.items()}
pd.DataFrame([{'파일': p.name, '바이트': p.stat().st_size} for p in [sif_path, *stat_paths.values()]])


프로젝트: /home/playdata/workspace/rag-project-1/src/rag_project_1/Project_1
저장소: /home/playdata/workspace/rag-project-1


,파일,바이트
0,한국산업안전보건공단_산업재해 고위험요인(SIF) 아카이브_20260401.xlsx,1431108
1,한국산업안전보건공단_산업중분류별 규모별 사고사망자수_20251231.csv,2219
2,한국산업안전보건공단_산업중분류별 규모별 사고재해자수_20251231.csv,2578
3,한국산업안전보건공단_산업중분류별 규모별 사망만인율_20251231.csv,2612
4,한국산업안전보건공단_산업중분류별 규모별 사업장수_20251231.csv,2774


## 2. 로드와 구조 확인

시트별 행 수, 실제 컬럼, 결측 수를 먼저 확인합니다. 건설업 첫 행은 `연번`이 없고 작업분류 헤더를 담은 보조 행입니다. CSV는 UTF-8 BOM을 허용합니다.


In [2]:
sif_sheets = {
    '제조업등': pd.read_excel(sif_path, sheet_name='아카이브(제조업등)', header=2),
    '건설업': pd.read_excel(sif_path, sheet_name='아카이브(건설업)', header=2),
}
stats = {name: pd.read_csv(path, encoding='utf-8-sig') for name, path in stat_paths.items()}

for name, frame in [*sif_sheets.items(), *stats.items()]:
    print(f'[{name}] {frame.shape[0]}행 × {frame.shape[1]}열')
    print('컬럼:', frame.columns.tolist())
    print('결측:', frame.isna().sum().to_dict())


[제조업등] 2573행 × 10열
컬럼: ['Unnamed: 0', '연번', '산재업종\n(대분류)', '산재업종\n(중분류)', '산재업종\n(소분류)', '재해개요', '기인물', '고위험작업·상황', '재해유발요인', '위험성 감소대책(예시)']
결측: {'Unnamed: 0': 2573, '연번': 0, '산재업종\n(대분류)': 0, '산재업종\n(중분류)': 0, '산재업종\n(소분류)': 0, '재해개요': 0, '기인물': 0, '고위험작업·상황': 0, '재해유발요인': 0, '위험성 감소대책(예시)': 0}
[건설업] 3460행 × 10열
컬럼: ['Unnamed: 0', '연번', '고위험작업·상황', 'Unnamed: 3', 'Unnamed: 4', '재해종류', '재해개요', '기인물', '재해유발요인', '위험성 감소대책(예시)']
결측: {'Unnamed: 0': 3460, '연번': 1, '고위험작업·상황': 0, 'Unnamed: 3': 0, 'Unnamed: 4': 0, '재해종류': 1, '재해개요': 1, '기인물': 1, '재해유발요인': 1, '위험성 감소대책(예시)': 1}
[사고사망자수] 30행 × 12열
컬럼: ['대업종', '구분', '5인 미만 근로자수', '5인 - 9인 근로자수', '10인 - 19인 근로자수', '20인 - 29인 근로자수', '30인 - 49인 근로자수', '50인 - 99인 근로자수', '100인 - 299인 근로자수', '300인 - 499인 근로자수', '500인 - 999인 근로자수', '1000인 이상 근로자수']
결측: {'대업종': 0, '구분': 0, '5인 미만 근로자수': 0, '5인 - 9인 근로자수': 0, '10인 - 19인 근로자수': 0, '20인 - 29인 근로자수': 0, '30인 - 49인 근로자수': 0, '50인 - 99인 근로자수': 0, '100인 - 299인 근로자수': 0, '300인 - 499인 근로자수': 0, '500인 - 999인 근로자수

## 3. 필요한 컬럼 정리와 결측·중복 처리

`page_content`는 임베딩하여 의미 검색할 자연어 본문입니다. `metadata`는 출처, 업종, 작업, 재해종류 등 필터와 출처 표시에 쓸 구조화 정보입니다. `source_id`는 파일·시트·연번 또는 CSV 행번호로 만든 안정적인 키입니다.

필수 내용(`재해개요`와 `고위험작업·상황` 중 하나)이 모두 비면 제외합니다. 선택 내용이 비면 해당 줄을 생략합니다. 동일 시트에서 주요 내용이 같은 중복 행은 하나만 남깁니다. CSV 수치 결측은 0으로 바꾸지 않고 **자료 없음**으로 적어 오해를 막습니다.


In [3]:
def clean(value):
    if pd.isna(value):
        return ''
    return re.sub(r'\s+', ' ', str(value)).strip()

def labeled_content(fields):
    return '\n'.join(f'{label}: {value}' for label, value in fields if value)

def make_doc(source_id, kind, content, metadata):
    return {'source_id': source_id, 'kind': kind, 'page_content': content, 'metadata': metadata}

docs = []
quality = []
for group, raw in sif_sheets.items():
    frame = raw.copy()
    before = len(frame)
    frame = frame[frame['연번'].notna()].copy()  # 건설업의 보조 헤더 제외
    frame['연번'] = frame['연번'].astype(int)
    if group == '제조업등':
        fields = ['재해개요', '기인물', '고위험작업·상황', '재해유발요인', '위험성 감소대책(예시)']
        meta_fields = {'대분류': '산재업종\n(대분류)', '중분류': '산재업종\n(중분류)', '소분류': '산재업종\n(소분류)'}
    else:
        fields = ['재해개요', '기인물', '고위험작업·상황', '재해유발요인', '위험성 감소대책(예시)']
        meta_fields = {'작업중분류': 'Unnamed: 3', '작업소분류': 'Unnamed: 4', '재해종류': '재해종류'}
    required = ['연번', *fields, *meta_fields.values()]
    missing = set(required) - set(frame.columns)
    if missing:
        raise ValueError(f'{group}에 필요한 컬럼이 없습니다: {missing}')
    frame = frame.drop_duplicates(subset=['연번'], keep='first')
    frame['_dedupe'] = frame[fields].apply(lambda row: '|'.join(clean(v) for v in row), axis=1)
    frame = frame.drop_duplicates(subset=['_dedupe'], keep='first')
    frame = frame[(frame['재해개요'].map(clean) != '') | (frame['고위험작업·상황'].map(clean) != '')]
    for _, row in frame.iterrows():
        metadata = {'source_file': sif_path.name, 'sheet': f'아카이브({group})', 'row_number': int(row['연번']), 'group': group}
        metadata.update({key: clean(row[col]) for key, col in meta_fields.items() if clean(row[col])})
        content = labeled_content([(field, clean(row[field])) for field in fields])
        docs.append(make_doc(f'sif:{group}:{int(row["연번"])}', 'sif_case', content, metadata))
    quality.append({'자료': group, '원본행': before, '문서': len(frame), '제외행': before - len(frame)})

pd.DataFrame(quality)


,자료,원본행,문서,제외행
0,제조업등,2573,2573,0
1,건설업,3460,3459,1


### 통계 CSV 4개를 별도 문서로 만들기

통계는 사고사례가 아니므로 `kind='industry_stat'`으로 구분합니다. 업종 한 행의 10개 규모별 값을 한 문서에 넣습니다. `대업종`과 `구분`은 metadata에도 저장합니다. 원본 0은 그대로 보존하고 결측은 `자료 없음`으로 표현합니다.


In [4]:
for metric, raw in stats.items():
    if not {'대업종', '구분'}.issubset(raw.columns):
        raise ValueError(f'{metric}: 대업종/구분 컬럼이 없습니다.')
    value_columns = [col for col in raw.columns if col not in ('대업종', '구분')]
    if len(value_columns) != 10:
        raise ValueError(f'{metric}: 규모별 컬럼이 10개가 아닙니다: {value_columns}')
    frame = raw.dropna(subset=['구분']).copy()
    frame = frame.drop_duplicates(subset=['대업종', '구분'], keep='first')
    for index, row in frame.iterrows():
        industry = clean(row['구분'])
        sector = clean(row['대업종'])
        values = [(str(col), '자료 없음' if pd.isna(row[col]) else clean(row[col])) for col in value_columns]
        content = labeled_content([('지표', metric), ('대업종', sector), ('구분', industry), *values])
        metadata = {'source_file': stat_paths[metric].name, 'row_number': int(index) + 2,
                    'metric': metric, 'industry': industry, 'sector': sector, 'reference_date': '2025-12-31'}
        docs.append(make_doc(f'stat:{metric}:{index}', 'industry_stat', content, metadata))
    quality.append({'자료': metric, '원본행': len(raw), '문서': len(frame), '제외행': len(raw) - len(frame)})

assert len({doc['source_id'] for doc in docs}) == len(docs), 'source_id 중복'
assert all(doc['page_content'] for doc in docs), '빈 page_content'
print(pd.DataFrame(quality).to_string(index=False))
print('전체 문서:', len(docs))
print('사례 예시:', docs[0])
print('통계 예시:', next(doc for doc in docs if doc['kind'] == 'industry_stat'))


    자료  원본행   문서  제외행
  제조업등 2573 2573    0
   건설업 3460 3459    1
사고사망자수   30   30    0
사고재해자수   30   30    0
 사망만인율   30   30    0
  사업장수   30   30    0
전체 문서: 6152
사례 예시: {'source_id': 'sif:제조업등:1', 'kind': 'sif_case', 'page_content': '재해개요: 2019년 10월경 OO사업장에서 재해자가 화물용 리프트에 제품을 싣는 작업 중 리프트와 작업장 사이의 개구부로 떨어져 사망\n기인물: 분류 불가\n고위험작업·상황: 리프트·승강기 관련 작업\n재해유발요인: 『화물용 승강기 또는 리프트(기계식 주차설비 포함)의 운반구(기)에서 작업 또는 이동』하다가 승강로 내부 구조물과 운반구 사이에 끼임 또는 그 사이 개구부로 떨어짐 위험\n위험성 감소대책(예시): ▶ 개구부에 추락방지용 덮개 설치 ▶ 카(운반구)에 울 추가 설치 ▶ PLC 및 각 층 위치센서를 설치하여 정해진 위치에 운반구가 정지하도록 설치', 'metadata': {'source_file': '한국산업안전보건공단_산업재해 고위험요인(SIF) 아카이브_20260401.xlsx', 'sheet': '아카이브(제조업등)', 'row_number': 1, 'group': '제조업등', '대분류': '제조업', '중분류': '화학및고무제품제조업', '소분류': '화학비료제조업'}}
통계 예시: {'source_id': 'stat:사고사망자수:0', 'kind': 'industry_stat', 'page_content': '지표: 사고사망자수\n대업종: 금융및보험업\n구분: 금융및보험업\n5인 미만 근로자수: 0\n5인 - 9인 근로자수: 0\n10인 - 19인 근로자수: 0\n20인 - 29인 근로자수: 0\n30인 - 49인 근로자수: 0\n50인 - 99인 근로자수: 0\n100인 - 299인 근로자수: 0\n300인 - 499인 

## 4. 임베딩 서비스 연결

임베딩은 문장을 숫자 벡터로 바꾸는 작업입니다. 이 WSL 프로젝트에 설치된 `langchain-openai`의 `OpenAIEmbeddings(model='text-embedding-3-small')`을 사용하며, 기본 차원 1536을 검증합니다. 처음에는 각 자료 유형을 포함한 20개만 API로 확인한 뒤 전체 적재를 선택하세요. 모델이나 `dimensions` 옵션을 바꾸면 DB의 벡터 차원도 함께 바꿔야 합니다.


In [5]:
from langchain_openai import OpenAIEmbeddings

MAX_DOCUMENTS = 20  # 전체 데이터 적재 시 None
BATCH_SIZE = 50
EMBEDDING_MODEL = 'text-embedding-3-small'
VECTOR_DIM = 1536

if not os.getenv('OPENAI_API_KEY'):
    raise EnvironmentError(f'{REPO_ROOT / ".env"}에 OPENAI_API_KEY를 설정하세요.')
def representative_sample(items, limit):
    # 제조업등·건설업 사례와 통계 CSV 네 종류를 순서대로 돌아가며 뽑습니다.
    buckets = {}
    for doc in items:
        key = (doc['kind'], doc['metadata'].get('group', doc['metadata'].get('metric')))
        buckets.setdefault(key, []).append(doc)
    selected = []
    while len(selected) < limit and any(buckets.values()):
        for bucket in buckets.values():
            if bucket and len(selected) < limit:
                selected.append(bucket.pop(0))
    return selected

selected_docs = docs if MAX_DOCUMENTS is None else representative_sample(docs, MAX_DOCUMENTS)
if not selected_docs:
    raise ValueError('적재할 문서가 없습니다.')
embedder = OpenAIEmbeddings(model=EMBEDDING_MODEL)
print(f'임베딩 대상 {len(selected_docs)} / 전체 {len(docs)} 문서')


임베딩 대상 20 / 전체 6152 문서


## 5. PostgreSQL/pgvector 테이블 생성 및 적재

`rag_day1_documents`는 기존 실습의 `documents` 테이블과 분리합니다. `source_id`를 기본키로 사용해 반복 실행해도 같은 자료가 중복되지 않습니다. 각 배치마다 임베딩을 만들고 `ON CONFLICT`로 갱신합니다. 중간에 실패하면 이미 끝난 배치는 남으므로 다시 실행하면 이어서 갱신할 수 있습니다.

PostgreSQL 서버와 API 키가 준비되지 않았다면 여기서 오류가 납니다. 위의 데이터 확인·전처리 셀은 DB 없이도 실행할 수 있습니다.


In [6]:
import numpy as np
import psycopg
from pgvector.psycopg import register_vector
from psycopg.types.json import Json

DB_URL = os.getenv('DATABASE_URL') or 'postgresql://postgres:pgvector-demo@127.0.0.1:5432/rag'  # docker-compose.yaml

def prepare_database(conn):
    conn.execute('CREATE EXTENSION IF NOT EXISTS vector')
    conn.execute('''
        CREATE TABLE IF NOT EXISTS rag_day1_documents (
            source_id text PRIMARY KEY,
            kind text NOT NULL,
            content text NOT NULL,
            metadata jsonb NOT NULL,
            embedding vector(1536) NOT NULL,
            embedding_model text NOT NULL,
            updated_at timestamptz NOT NULL DEFAULT now()
        )
    ''')
    conn.commit()
    register_vector(conn)

def save_batch(conn, batch, vectors):
    if len(batch) != len(vectors) or any(len(v) != VECTOR_DIM for v in vectors):
        raise ValueError('문서 수 또는 임베딩 차원이 맞지 않습니다.')
    rows = [(doc['source_id'], doc['kind'], doc['page_content'], Json(doc['metadata']), np.asarray(vector, dtype=np.float32), EMBEDDING_MODEL)
            for doc, vector in zip(batch, vectors)]
    with conn.cursor() as cur:
        cur.executemany('''
            INSERT INTO rag_day1_documents
                (source_id, kind, content, metadata, embedding, embedding_model)
            VALUES (%s, %s, %s, %s, %s, %s)
            ON CONFLICT (source_id) DO UPDATE SET
                kind = EXCLUDED.kind,
                content = EXCLUDED.content,
                metadata = EXCLUDED.metadata,
                embedding = EXCLUDED.embedding,
                embedding_model = EXCLUDED.embedding_model,
                updated_at = now()
        ''', rows)
    conn.commit()

with psycopg.connect(DB_URL, connect_timeout=5) as conn:
    prepare_database(conn)
    for start in range(0, len(selected_docs), BATCH_SIZE):
        batch = selected_docs[start:start + BATCH_SIZE]
        vectors = embedder.embed_documents([doc['page_content'] for doc in batch])
        save_batch(conn, batch, vectors)
        print(f'{start + len(batch)} / {len(selected_docs)} 적재')


20 / 20 적재


## 6. 적재 검증 및 간단 검색

건수, 유형별 건수, 벡터 차원과 본문 일부를 확인합니다. 테이블 전체 건수는 이전 시험 실행에서 남은 행 때문에 이번 선택 문서 수보다 클 수 있습니다. 마지막에는 질문을 임베딩하여 pgvector의 코사인 거리(`<=>`)로 가까운 문서 3개를 조회합니다. 여기까지가 Day 1의 검색 준비 완료 기준입니다.


In [7]:
with psycopg.connect(DB_URL, connect_timeout=5) as conn:
    register_vector(conn)
    total = conn.execute('SELECT count(*) FROM rag_day1_documents').fetchone()[0]
    by_kind = conn.execute('SELECT kind, count(*) FROM rag_day1_documents GROUP BY kind ORDER BY kind').fetchall()
    sample = conn.execute('''
        SELECT source_id, kind, left(content, 120), vector_dims(embedding)
        FROM rag_day1_documents ORDER BY source_id LIMIT 3
    ''').fetchall()
    print('테이블 전체 건수:', total, '/ 이번 실행 대상:', len(selected_docs))
    print('유형별 건수:', by_kind)
    print('샘플 (ID, 유형, 본문 일부, 차원):', sample)
    ids = [doc['source_id'] for doc in selected_docs]
    present = conn.execute('SELECT count(*) FROM rag_day1_documents WHERE source_id = ANY(%s)', (ids,)).fetchone()[0]
    assert present == len(selected_docs), f'이번 실행 문서 중 {len(selected_docs) - present}개 누락'
    assert all(row[3] == VECTOR_DIM for row in sample), '임베딩 차원이 다릅니다.'

    question = '화물용 리프트 작업 중 추락을 막는 대책은?'
    query_vector = embedder.embed_query(question)
    results = conn.execute('''
        SELECT source_id, kind, metadata, left(content, 180), embedding <=> %s AS distance
        FROM rag_day1_documents
        WHERE source_id = ANY(%s) AND embedding_model = %s
        ORDER BY distance LIMIT 3
    ''', (np.asarray(query_vector, dtype=np.float32), ids, EMBEDDING_MODEL)).fetchall()
    for source_id, kind, metadata, excerpt, distance in results:
        print(f'[{source_id}] {kind}, 거리={distance:.4f}, 출처={metadata["source_file"]}')
        print(excerpt, '\n')


테이블 전체 건수: 36 / 이번 실행 대상: 20
유형별 건수: [('industry_stat', 12), ('sif_case', 24)]
샘플 (ID, 유형, 본문 일부, 차원): [('sif:제조업등:1', 'sif_case', '재해개요: 2019년 10월경 OO사업장에서 재해자가 화물용 리프트에 제품을 싣는 작업 중 리프트와 작업장 사이의 개구부로 떨어져 사망\n기인물: 분류 불가\n고위험작업·상황: 리프트·승강기 관련 작업\n재해유발요인: 『', 1536), ('sif:건설업:1', 'sif_case', '재해개요: 2019년 03월경 ○○ 현장 1층에서 피재자가 장비(미니 백호 1ton)진입을 위하여 해당 방호시설(나무합판, 석자재, 고체연료통)을 이동시키다가 개구부에 빠져 약 4.7m 아래인 지하층 바닥으로 추락하', 1536), ('sif:제조업등:10', 'sif_case', '재해개요: 2017년 3월경 OO사업장 창고에서 재해자가 이동 중 후진하던 페이로더에 깔려 사망\n기인물: 건설기계(로더)\n고위험작업·상황: 작업장소 통행·이동\n재해유발요인: 『차량(하역운반기계 또는 건설기계 포함)작', 1536)]
[sif:제조업등:1] sif_case, 거리=0.5715, 출처=한국산업안전보건공단_산업재해 고위험요인(SIF) 아카이브_20260401.xlsx
재해개요: 2019년 10월경 OO사업장에서 재해자가 화물용 리프트에 제품을 싣는 작업 중 리프트와 작업장 사이의 개구부로 떨어져 사망
기인물: 분류 불가
고위험작업·상황: 리프트·승강기 관련 작업
재해유발요인: 『화물용 승강기 또는 리프트(기계식 주차설비 포함)의 운반구(기)에서 작업 또는 이동』하다가 승강로 내부 구조 

[sif:건설업:1] sif_case, 거리=0.6698, 출처=한국산업안전보건공단_산업재해 고위험요인(SIF) 아카이브_20260401.xlsx
재해개요: 2019년 03월경 ○○ 현장 1층에서 피재자가 장비(미니 백호 1ton)진입을 위하여 해당 방호시설(나무합판,

## 완료 확인과 다음 학습

- 로드 결과에서 SIF 제조업등 2,573행, 건설업 3,460행(보조 헤더 1행 포함), CSV 4개 각각 30행인지 확인합니다.
- `quality`에서 제외 행이 왜 생겼는지 확인합니다. 원본 파일은 수정하지 않습니다.
- 적재 검증의 `present`가 이번 실행 대상 수와 같고, 벡터 차원이 1536이면 완료입니다.
- 짧은 변형 실습: `MAX_DOCUMENTS = 20`을 유지한 채 `docs`에서 `group == '건설업'` 사례 20개를 선택해 다시 적재하고, 검색 질문을 바꿔 결과의 `sheet`와 본문을 비교해 보세요. 전체 6,152개 문서를 적재할 때는 `MAX_DOCUMENTS = None`으로 바꿉니다. 그 전에 API 비용과 처리 시간을 확인하세요.

**범위**: 다음 날 Retriever와 LLM 답변을 연결합니다. 이 노트북은 Agent나 화면 UI를 만들지 않습니다.
